master/buildbot/wamp/connector.py

```diff
@@ -0,0 +1,140 @@
+        log.msg("We don't know how to recover this without restarting the whole system")
+        log.msg(str(details))
+        yield self.master.stopService()
+
+    def onUserError(self, e, msg):
+        log.err(e, msg)
+
+
+def make(config):
+    if config:
+        return MasterService(config)
+    else:
+        # if no config given, return a description of this WAMPlet ..
+        return {'label': 'Buildbot master wamplet',
+                'description': 'This contains all the wamp methods provided by a buildbot master'}
+
+
+class WampConnector(service.ReconfigurableServiceMixin, service.AsyncMultiService):
+    serviceClass = Service
+
+    def __init__(self, master):
+        service.AsyncMultiService.__init__(self)
+        self.setName('wamp')
+        self.master = master
+        self.app = self.router_url = None
+        self.serviceDeferred = defer.Deferred()
+        self.service = None
+
+    def getService(self):
+        if self.service is not None:
+            return defer.succeed(self.service)
+        d = defer.Deferred()
+
+        @self.serviceDeferred.addCallback
+        def gotService(service):
+            d.callback(service)
+            return service
+        return d
+
+    @defer.inlineCallbacks
+    def publish(self, topic, data, options=None):
+        service = yield self.getService()
+        try:
+            ret = yield service.publish(topic, data, options=options)
+        except TransportLost:
+            log.err(failure.Failure(), "while publishing event " + topic)
+            return
+        defer.returnValue(ret)
+
+    @defer.inlineCallbacks
+    def subscribe(self, callback, topic=None, options=None):
+        service = yield self.getService()
+        ret = yield service.subscribe(callback, topic, options)
+        defer.returnValue(ret)
+
+    def reconfigServiceWithBuildbotConfig(self, new_config):
+        if new_config.mq.get('type', 'simple') != "wamp":
+            return
+        wamp = new_config.mq
+        log.msg("Starting wamp with config: %r", wamp)
+        router_url = wamp.get('router_url', None)
+
+        # This is not a good idea to allow people to switch the router via reconfig
+        # how would we continue the current transactions ?
+        # how would we tell the slaves to switch router ?
+        if self.app is not None and self.router_url != router_url:
+            raise ValueError("Cannot use different wamp router url when reconfiguring")
+        if router_url is None:
+            return
+        self.router_url = router_url
+        self.app = self.serviceClass(
+            url=self.router_url,
+            extra=dict(master=self.master, parent=self),
+            realm=wamp.get('realm', 'buildbot'),
+            make=make,
+            debug=wamp.get('debug_websockets', False),
+            debug_wamp=wamp.get('debug_lowlevel', False),
+            debug_app=wamp.get('debug', False)
+        )
+
+        self.app.setServiceParent(self)
```

Comment:
1. this will not trigger a stopService/startService ?
2. shouldn't you yield it ? (I'm kind of lost with the overall API for
startService
stopService
setServiceParent
disownParent
as some service Class return deferreds, and others don't.